Upload timed text in the frontend

mbenadda opened the pull request:
## Purpose

Users need to be able to create new subtitles/closed captions/transcript tracks to go along with their videos and manage those tracks.

## Proposal

Add a pane to the Dashboard that will include all the necessary UI elements to take care of timed text tracks.

- [x] Make the `<DashboardVideoPane />` more compact to make room for the timed text tracks pane
- [x] Add an action for resource deletion in the `resourceById` reducer
- [x] Implement the `<DashboardTimedTextPane />` along with its subparts and related side-effects.

lunika commented on src/frontend/data/sideEffects/createTimedTextTrack/createTimedTextTrack.ts:
```diff
@@ -0,0 +1,38 @@
+    {
+      body: JSON.stringify({ language, mode }),
+      headers: {
+        Authorization: `Bearer ${jwt}`,
```
you should maybe throw an error if `jwt` is null ? The API will reject you and you save a call to the server.

lunika commented on src/frontend/data/sideEffects/deleteTimedTextTrack/deleteTimedTextTrack.ts:
```diff
@@ -0,0 +1,32 @@
+    `${API_ENDPOINT}/${modelName.TIMEDTEXTTRACKS}/${timedtexttrack.id}/`,
+    {
+      headers: {
+        Authorization: `Bearer ${jwt}`,
```
same here, you can throw an error earlier if `jwt` is null ?

lunika commented on src/frontend/data/sideEffects/getTimedTextTrackLanguageChoices/getTimedTextTrackLanguageChoices.ts:
```diff
@@ -0,0 +1,38 @@
+    `${API_ENDPOINT}/${modelName.TIMEDTEXTTRACKS}/`,
+    {
+      headers: {
+        Authorization: `Bearer ${jwt}`,
```
Same here with the `jwt` that can be `null`

lunika commented on src/frontend/data/sideEffects/getTimedTextTrackList/getTimedTextTrackList.ts:
```diff
@@ -0,0 +1,27 @@
+    `${API_ENDPOINT}/${modelName.TIMEDTEXTTRACKS}/`,
+    {
+      headers: {
+        Authorization: `Bearer ${jwt}`,
```
Same here with the `jwt` that can be null

lunika commented on src/frontend/data/sideEffects/initiateUpload/initiateUpload.ts:
```diff
@@ -0,0 +1,35 @@
+    `${API_ENDPOINT}/${objectType}/${objectId}/initiate-upload/`,
+    {
+      headers: {
+        Authorization: `Bearer ${jwt}`,
```
Same here with the `jwt` that can be `null`

lunika commented on src/frontend/tslint.json:
```diff
@@ -5,6 +5,7 @@
   "rules": {
     "arrow-parens": false,
     "interface-name": [true, "never-prefix"],
+    "max-line-length": false,
```
Are you sure you really want to disable it ? Why did you choose to do that ? Why not allowing a higher length ?

mbenadda replied:
As we use `prettier`, having this rule in `tslint` is 99% pointless: `prettier` will break down long lines for us (and the prettier check will ensure we ran it).

As for the 1%: there is only one case where `prettier` does not break down a long line (a very long import with a single imported item). As a general policy I think it's best to prioritize `prettier` over `tslint`, and so that's what we're doing by disabling this check.

mbenadda commented on src/frontend/components/TimedTextCreationForm/TimedTextCreationForm.tsx:
```diff
@@ -0,0 +1,163 @@
+/** State shape for the TimedTextCreationForm component. */
+interface TimedTextCreationFormState {
+  error?: 'creation' | 'schema';
+  languageChoices: SelectOption[];
```
I think it is the responsibility of this component to retrieve this list.

Sure, we can optimize it by moving the call up in the component tree, but in my opinion the better way to do this is to add caching for the call itself.

Since the side effects are to be replaced with a cleaner structure through `saga` or `thunks` I chose not to implement that in this PR.

mbenadda commented on src/frontend/components/TimedTextListItem/TimedTextListItem.tsx:
```diff
@@ -0,0 +1,117 @@
+  }
+
+  async componentDidMount() {
+    const languageChoices = await getTimedTextTrackLanguageChoices(
```
I'd use the same reasoning as with `TimedTextCreationForm`. The component that needs it should be the one to request it.

mbenadda commented on src/frontend/components/TimedTextCreationForm/TimedTextCreationForm.tsx:
```diff
@@ -0,0 +1,163 @@
+      <TimedTextCreationFormStyled>
+        <FormattedMessage {...messages.addTrackLabel} />
+        <Select
+          onChange={this.onSelectChange.bind(this)}
```
IMO binding in the constructor is more confusing and having to bind methods is one of the "bad parts" of react. I'd rather not do this until we have an actual problem with the render binding.

Hopefully Hooks are released soon and we can get rid of this problem altogether 😊

mbenadda replied:
This is a stage 3 proposal. At this point I think it will probably be implemented but I think it's a good policy to refrain from using stage 3 and below as much as possible.

mbenadda commented:
> In the Dropzone, you didn't manage the object type uploaded. So you always have Select a video to upload display.

Fixed 👍
I just replaced `video` with `file` as the kind of object (video, subtitle, transcript, cc) appears in the Heading above anyway.

> in the instructor view, a link to the dashboard should be added. Otherwise going to the dashboard is nor really easy.

Done 👌

> If I already have a video in ready state, in the dashboard I click on Replace the video then I'm redirected to the instructor view and I have to click again on the replace video at the bottom of the instructor view to reach the upload view.

Fixed 👍

> what do you think about installing the fetch polyfill ?

What do you have in mind? I think `fetch` is supported in all modern browsers. I'm not sure Marsha works at all in any browser that does not support `fetch`.

> you should maybe throw an error if jwt is null ? The API will reject you and you save a call to the server.
> [and related `jwt` comments]

I agree with you that there's an inaccuracy there. However, I think downstream at the API call sites is not the right place to fix it. We should try and get a better type upstream so when we reach these calls the type reflects the expected reality: `jwt` should not be `Nullable`.

Indeed, especially with the changes for xAPI, `jwt` is **always** supposed to be defined, except when the initial LTI request fails.
We should be able to reflect this in the `AppState` or at least in `RootState['context']` in the store.

I think I should try to fix this in another PR. WDYT?

mbenadda commented:
@lunika thanks for the review, I adressed/answered your comments 👌

lunika commented:
Lot of code will be replaced in this PR once `saga` or `thunk` will be added. What do you think about doing this jib in a new PR, we merge it and then you rework this one ?

lunika commented:
> I agree with you that there's an inaccuracy there. However, I think downstream at the API call sites is not the right place to fix it. We should try and get a better type upstream so when we reach these calls the type reflects the expected reality: jwt should not be Nullable.

> Indeed, especially with the changes for xAPI, jwt is always supposed to be defined, except when the initial LTI request fails.
We should be able to reflect this in the AppState or at least in RootState['context'] in the store.

> I think I should try to fix this in another PR. WDYT?

I agree, we can do this in an other PR. Can you please open an issue ? Thanks

mbenadda commented:
> Lot of code will be replaced in this PR once `saga` or `thunk` will be added. What do you think about doing this job in a new PR, we merge it and then you rework this one ?

I'd much rather merge this and do some refactoring later on. Long running PRs are a productivity killer for me and I don't think the changes related to side effects here are dealbreakers that prevent a merge.

mbenadda commented:
> I agree, we can do this in an other PR. Can you please open an issue ? Thanks

👍
#197
